File: 274. H-Index/main.cpp
#include <algorithm>
#include <vector>

using std::reverse;
using std::sort;
using std::vector;

class Solution {
public:
    int hIndex(vector<int> &citations) {
        vector<int> dp(ARRAY_SIZE + 1, 0);

        for (const auto &citation : citations) {
            dp[citation]++;
        }

        for (int i = ARRAY_SIZE - 1; i >= 0; i--) {
            dp[i] += dp[i + 1];
            if (dp[i] >= i) {
                return i;
            }
        }

        return 0;
    }

private:
    static constexpr int ARRAY_SIZE = 1001;
};

class Solution {
public:
    int hIndex(vector<int> &citations) {
        sort(citations.begin(), citations.end(), [](int a, int b) { return a > b; });

        int h = 0;
        for (int i = 0; i < citations.size(); i++) {
            if (citations[i] < i + 1) {
                break;
            }

            h = i + 1;
        }

        return h;
    }
};

class Solution {
public:
    int hIndex(vector<int> &citations) {
        sort(citations.begin(), citations.end());
        reverse(citations.begin(), citations.end());

        int h = 0;
        for (int i = 0; i < citations.size(); i++) {
            if (citations[i] < i + 1) {
                break;
            }

            h = i + 1;
        }

        return h;
    }
};

File: 274. H-Index/main.rs
struct Solution;

const ARRAY_SIZE: usize = 1001;

impl Solution {
    pub fn h_index(citations: Vec<i32>) -> i32 {
        let mut dp = vec![0; ARRAY_SIZE + 1];

        for citation in citations {
            dp[citation as usize] += 1;
        }

        for i in (0..ARRAY_SIZE - 1).rev() {
            dp[i] += dp[i + 1];
            if dp[i] >= i {
                return i as i32;
            }
        }

        0
    }
}

impl Solution {
    pub fn h_index(citations: Vec<i32>) -> i32 {
        let mut citations = citations;
        citations.sort_by(|a, b| b.cmp(a));

        let mut h = 0;
        for i in 0..citations.len() {
            if citations[i] < i as i32 + 1 {
                break;
            }

            h = i as i32 + 1;
        }

        h
    }
}

impl Solution {
    pub fn h_index(citations: Vec<i32>) -> i32 {
        let mut citations = citations;
        citations.sort();
        citations.reverse();

        let mut h = 0;
        for i in 0..citations.len() {
            if citations[i] < i as i32 + 1 {
                break;
            }

            h = i as i32 + 1;
        }

        h
    }
}
